### PR TITLE
don't use ext4 mkfs options for other fs

### LINKED ISSUE #16528: wrong options passed to mkfs if fstype != ext4
[here](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount.go#L86) the option is for ext4. If `fstype != 'ext4'` then `mkfs` will fail.

### PR DESCRIPTION
**Pull request by rootfs**
fix #16528

@kubernetes/rh-storage

### CODE REVIEW

**Comment by swagiaal** [comment]
@rootfs good catch... could you add that scenario to safe_format_and_mount_test.go ? Otherwise LGTM. If we decide to decorate other file systems with default options we can change that to a switch statement.

What are your thoughts on the options BTW I just copied them from the old safe_format_and_mount script ?

**Comment by rootfs** [comment]
@swagiaal unit test updated. 
I am not quite sure with ext3/4 options. They are not defaults. And they make mkfs slower. I am a bit concerned (but haven't verified) if mkfs over large disks takes too long 
```man
                   lazy_itable_init[= <0 to disable, 1 to enable>]
                          If enabled and the uninit_bg feature is enabled, the inode table will not be fully initialized by mke2fs.  This speeds up filesystem initial‐
                          ization  noticeably, but it requires the kernel to finish initializing the filesystem in the background when the filesystem is first mounted.
                          If the option value is omitted, it defaults to 1 to enable lazy inode table zeroing.

                   lazy_journal_init[= <0 to disable, 1 to enable>]
                          If enabled, the journal inode will not be fully zeroed out by mke2fs.  This speeds up filesystem initialization noticeably, but carries  some
                          small risk if the system crashes before the journal has been overwritten entirely one time.  If the option value is omitted, it defaults to 1
                          to enable lazy journal inode zeroing.
```

File: pkg/util/mount/safe_format_and_mount_test.go [hunk 1]
```diff
@@ -105,6 +105,44 @@ func TestSafeFormatAndMount(t *testing.T) {
...
+		},
+		{ // Test that 'lsblk' is called and confirms unformatted disk, format passes, second mount passes
+			fstype:    "xfs",
+			mountErrs: []error{fmt.Errorf("unknown filesystem type '(null)'"), nil},
+			execScripts: []ExecArgs{
+				{"lsblk", []string{"-nd", "-o", "FSTYPE", "/dev/foo"}, "", nil},
+				{"mkfs.xfs", []string{"/dev/foo"}, "", nil},
+			},
+			expectedError: nil,
+		},
```
Comment (swagiaal) [inline pkg/util/mount/safe_format_and_mount_test.go]: Oh we don't need all the test a different fstype with all the test cases. Just one to test that ext4 gets run with the flags (probably already there) and one to test that non ext4 gets run without the flag.

Comment (rootfs) [inline pkg/util/mount/safe_format_and_mount_test.go]: all tests there just test ext4 with flags. new tests test all cases of non-ext fs without flags.

Comment (swagiaal) [inline pkg/util/mount/safe_format_and_mount_test.go]: In fact just keep the last test case above and delete the rest. Change to comment to something like "test that none ext4 fs does not get called with ext4 options.

Comment (rootfs) [inline pkg/util/mount/safe_format_and_mount_test.go]: cool, updated

Comment (swagiaal) [inline pkg/util/mount/safe_format_and_mount_test.go]: > cool, updated

Thanks... LGTM

**Comment by ashcrow** [comment]
Shippable has been ```waiting``` for about 4 hours. I have a feeling it will timeout.

This looks good to me!

File: pkg/util/mount/mount.go [hunk 2]
```diff
@@ -78,12 +78,15 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
...
 		if err == nil && notFormatted {
+			args := []string{source}
 			// Disk is unformatted so format it.
 			// Use 'ext4' as the default
 			if len(fstype) == 0 {
 				fstype = "ext4"
 			}
-			args := []string{"-E", "lazy_itable_init=0,lazy_journal_init=0", "-F", source}
+			if fstype == "ext4" || fstype == "ext3" {
+				args = []string{"-E", "lazy_itable_init=0,lazy_journal_init=0", "-F", source}
```
Comment (thockin) [inline pkg/util/mount/mount.go]: shouldn't this all go in a *_linux .go file?

Comment (swagiaal) [inline pkg/util/mount/mount.go]: > shouldn't this all go in a *_linux .go file?

Ah yes they should. I filed https://github.com/kubernetes/kubernetes/issues/16729. Would you be okay with me moving them after this PR and #14200 ?

Comment (rootfs) [inline pkg/util/mount/mount.go]: sounds a plan for me, @thockin ?

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Pull request merged**